aten/src/ATen/native/cuda/TensorTransformations.cu

```diff
@@ -0,0 +1,121 @@
+#include "ATen/native/TensorTransformations.h"
+
+#include "ATen/cuda/detail/IndexUtils.cuh"
+#include "ATen/NativeFunctions.h"
+#include "ATen/cuda/CUDATensorMethods.cuh"
+#include "ATen/cuda/CUDATypeConversion.cuh"
+
+namespace at {
+namespace native {
+
+#define AT_APPLY_THREADS_PER_BLOCK 32 * 16
+#define AT_APPLY_BLOCKS_PER_SM 4
+
+template <typename scalar_t, typename IndexType>
+#if __CUDA_ARCH__ >= 350
+__launch_bounds__(AT_APPLY_THREADS_PER_BLOCK, AT_APPLY_BLOCKS_PER_SM)
+#endif
+__global__ void
+kernel_pointwise_flip_apply2(const cuda::detail::TensorInfo<scalar_t, IndexType> in_tensor_info,
+                          cuda::detail::TensorInfo<scalar_t, IndexType> out_tensor_info,
+                          IndexType N,
+                          int flip_dim,
+                          int64_t total_dims) {
+  for (IndexType linear_index = blockIdx.x * blockDim.x + threadIdx.x; linear_index < N; linear_index += gridDim.x * blockDim.x) {
+    int64_t cur_indices = linear_index, rem = 0, dst_offset = 0;
```

Comment:
Can you please check if using int64_t indices is hurting performance? Usually all THC kernels are templated on index type, and for the tensors that can be indexed by 32-bit int they use 32-bit integers. I suspect if you use int32_t where appropriate perf will be better.

Reply:
Yes, I can check that. But will `int32_t` limits the total number of elements in a tensor since `N` is a `int64_t`?

Reply:
It's int64_t because you hardcode it to int64_t, in THC pointwise kernels are dispatched either to int64_t variants or int32_t variants depending on the number of elements. If it does not affect the performance, using int64_t is certainly safer and saves you from another template instantiation, however, I suspect it does affect performance.

Reply:
Ah, I see! Will look into it! Thanks for the super clear clarification 👍

Reply:
You are right! `int64_t` does hurt the performance. For case:

```
data_cuda = torch.arange(1000000, device=cuda).view(1000,1000)
%timeit data_cuda.flip(0)
```

Results look like these:

  | time
-- | --
hardcode int64_t | 178 µs
hardcode int32_t | 116 µs
templating int64_t | 139 µs
templating int32_t | 102 µs

Maybe templating int64_t is the way to go?

Reply:
I'm really at a loss why templating would produce different results than hardcoding, but I could live with templated int64_t numbers, until flip really becomes someone's bottleneck.

Reply:
@weiyangfb you need to insert `torch.cuda.synchronize()` calls before and after the relevant code, if you want accurate benchmarking.

For example:

```
%timeit torch.cuda.synchronize(); data_cuda.flip(0); torch.cuda.synchronize()
```

Reply:
@soumith Thanks a lot! The results of hardcode and templating are the same now:

  | time
-- | --
hardcode int64_t | 165 µs
hardcode int32_t | 128 µs
templating int64_t | 165 µs
templating int32_t | 128 µs

@ngimel will it be fine to template on index type = int64_t in this case?

Reply:
To me it's still a significant perf difference, but on the other hand if flip is a small fraction of the workload no one notice it. Up to @soumith to decide if it's worth additional code complexity and binary size.

Reply:
i vote for single template, no need to specialize two cases. it's indeed a niche function for now

Reply:
Sounds good! I will keep the template.